tests/lean/run/decideTactic.lean

```diff
@@ -0,0 +1,102 @@
+/-!
+# Tests of the `decide` tactic
+-/
+
+/-!
+Success
+-/
+#guard_msgs in
+example : 2 + 2 ≠ 5 := by decide
+
+
+/-!
+False proposition
+-/
+
+/--
+error: tactic 'decide' proved that the proposition
+  1 ≠ 1
+is false
+-/
+#guard_msgs in
+example : 1 ≠ 1 := by decide
+
+
+/-!
+Irreducible decidable instance
+-/
+opaque unknownProp : Prop
+
+/--
+error: tactic 'decide' failed for proposition
+  unknownProp
+since its Decidable instance
+  Classical.propDecidable unknownProp
+did not reduce to isTrue or isFalse.
+
+After unfolding the instance Classical.propDecidable, reduction got stuck at the Decidable instance
+  Classical.choice ⋯
+
+Hint: Reduction got stuck on Classical.choice, suggesting that there are "classical" instances, due
+to for example an 'open scoped Classical' command. The 'decide' tactic works by evaluating a
+decision procedure via reduction, but such classical instances cannot be evaluated.
+-/
+#guard_msgs in
+open scoped Classical in
+example : unknownProp := by decide
+
+
+/-!
+Reporting unfolded instances and give hint about Eq.rec.
+From discussion with Heather Macbeth on Zulip
+-/
+structure Nice (n : Nat) : Prop :=
+  (large : 100 ≤ n)
+
+theorem nice_iff (n : Nat) : Nice n ↔ 100 ≤ n := ⟨Nice.rec id, Nice.mk⟩
+
+def baz (n : Nat) : Decidable (Nice n) := by
+  rw [nice_iff]
+  infer_instance
+
+instance : Decidable (Nice n) := baz n
+
+/--
+error: tactic 'decide' failed for proposition
+  Nice 102
+since its Decidable instance
+  instDecidableNice
+did not reduce to isTrue or isFalse.
+
+After unfolding the instances baz and instDecidableNice, reduction got stuck at the Decidable instance
+  ⋯ ▸ inferInstance
+
+Hint: Reduction got stuck on '▸' (Eq.rec), which suggests that one of the decidability instances is
+defined using tactics such as 'rw' or 'simp'. Use definitions such as inferInstanceAs or
+decidable_of_decidable_of_iff to alter a proposition.
```

Comment:
```suggestion
Hint: Reduction got stuck on '▸' (Eq.rec), which suggests that one of the 'Decidable' instances is
defined using tactics such as 'rw' or 'simp'. Instead of these tactics, use functions such as 'inferInstanceAs' or
'decidable_of_decidable_of_iff' to alter a proposition.
```
I'm not too happy with the term "function" here, but I think it's better than "definitions" (as the definition is not what's being used, really). Perhaps "operator"?

Thanks for being careful about making them all clickable :)

Reply:
For me, "declaration" and "definition" have the same downside: a declaration/definition is that which results in a name being given a meaning, but not the name nor the meaning itself. User's aren't expected to stick `def ...` here, they're expected to invoke something that was previously defined.

This is indeed a bit nitpicky, and I think the risk of someone being misled is fairly low.

"function", then?

Reply:
Imagine if these were theorems instead of definitions. Would you have the same objection to "Instead of these tactics, use theorems such as 'thm1' or 'thm2'"? Definitions and theorems are fundamental notions that we need to refer to somehow (both of which are kind of declarations).

Is the analogy "definition is to function/constant as theorem is to theorem"?